algorithms/src/snark/marlin/ahp/matrices.rs

```diff
@@ -40,21 +37,37 @@ pub(crate) fn to_matrix_helper<F: Field>(matrix: &[Vec<(F, VarIndex)>], num_inpu
     cfg_iter!(matrix)
         .map(|row| {
             let mut row_map = BTreeMap::new();
-            row.iter().for_each(|(fe, column)| {
-                let column = match column {
-                    VarIndex::Public(i) => *i,
-                    VarIndex::Private(i) => num_input_variables + i,
-                };
-                *row_map.entry(column).or_insert_with(F::zero) += *fe;
+            row.iter().for_each(|(val, column)| {
+                if *val != F::zero() {
+                    let column = match column {
+                        VarIndex::Public(i) => *i,
+                        VarIndex::Private(i) => num_input_variables + i,
+                    };
+                    *row_map.entry(column).or_insert_with(F::zero) += *val;
+                }
             });
-            row_map.into_iter().map(|(column, coeff)| (coeff, column)).collect()
+            row_map.into_iter().map(|(column, val)| (val, column)).collect()
         })
         .collect()
 }
 
-/// This must *always* be in sync with `make_matrices_square`.
-pub(crate) fn padded_matrix_dim(num_formatted_variables: usize, num_constraints: usize) -> usize {
-    core::cmp::max(num_formatted_variables, num_constraints)
+/// Adds variables to randomize each z_M and preserve zero-knowledge
+pub(crate) fn add_randomizing_variables<F: PrimeField, CS: ConstraintSystem<F>>(
+    cs: &mut CS,
+    rand_assignments: &[F],
+    circuit_id: Option<CircuitId>,
+) {
+    // TODO: pass 3-sized array
+    let mut zk_vars = Vec::with_capacity(3);
+    for i in 0..3 {
+        zk_vars.push(cs.alloc(|| format!("random_{i}"), || Ok(rand_assignments[i])).unwrap());
+    }
+    cs.enforce(
+        || format!("constraint zk {:?}", circuit_id),
+        |lc| lc + zk_vars[0],
+        |lc| lc + zk_vars[1],
+        |lc| lc + zk_vars[2],
```

Comment:
```suggestion
        |lc| lc + zk_vars.next().unwrap(),
        |lc| lc + zk_vars.next().unwrap(),
        |lc| lc + zk_vars.next().unwrap(),
```

Reply:
Unfortunately: `cannot borrow as mutable because previous closure requires unique access
second borrow occurs here`. The three closures all want to capture `zk_vars`. I think the overhead of the original is minimal.